File: python3/bron_kerbosch1b.py
# coding: utf-8

from graph import UndirectedGraph, Vertex
from consumer import CliqueConsumer
from typing import List, Set


def explore(graph: UndirectedGraph, consumer: CliqueConsumer) -> None:
    """Naive Bron-Kerbosch algorithm, optimized"""
    if candidates := graph.connected_vertices():
        visit(
            graph=graph,
            consumer=consumer,
            candidates=candidates,
            excluded=set(),
            clique=[],
        )


def visit(
    graph: UndirectedGraph,
    consumer: CliqueConsumer,
    candidates: Set[Vertex],
    excluded: Set[Vertex],
    clique: List[Vertex],
) -> None:
    assert all(graph.has_degree(v) for v in candidates)
    assert all(graph.has_degree(v) for v in excluded)
    assert candidates.isdisjoint(excluded)
    assert candidates
    while candidates:
        v = candidates.pop()
        neighbours = graph.adjacencies[v]
        neighbouring_candidates = candidates.intersection(neighbours)
        if neighbouring_candidates:
            neighbouring_excluded = excluded.intersection(neighbours)
            visit(
                graph,
                consumer,
                candidates=neighbouring_candidates,
                excluded=neighbouring_excluded,
                clique=clique + [v],
            )
        elif excluded.isdisjoint(neighbours):
            consumer.accept(clique + [v])
        excluded.add(v)

File: rust/bron_kerbosch/src/core/bron_kerbosch1b.rs
//! Naive Bron-Kerbosch algorithm, optimized

use super::clique_consumer::CliqueConsumer;
use super::pile::Pile;
use crate::{BronKerboschAlgorithm, CliqueAccumulator, Graph, Vertex, VertexSetLike};
use std::ops::Not;

type CliqueInProgress<'a> = Pile<'a, Vertex>;

pub struct Algo();
impl BronKerboschAlgorithm for Algo {
    fn name() -> String {
        String::from("Ver1½")
    }

    fn explore<VertexSet, Accumulator>(
        graph: &Graph<VertexSet>,
        min_clique_size: usize,
        mut accumulator: Accumulator,
    ) -> Accumulator::Harvest
    where
        VertexSet: VertexSetLike,
        Accumulator: CliqueAccumulator,
    {
        let candidates: VertexSet = graph.connected_vertices().collect();
        let num_candidates = candidates.len();
        if num_candidates > 0 {
            let mut consumer = CliqueConsumer {
                min_clique_size,
                accu: &mut accumulator,
            };
            visit(
                graph,
                &mut consumer,
                candidates,
                VertexSet::with_capacity(num_candidates),
                &Pile::EMPTY,
            );
        }
        accumulator.harvest()
    }
}

fn visit<VertexSet, Accumulator>(
    graph: &Graph<VertexSet>,
    consumer: &mut CliqueConsumer<Accumulator>,
    mut candidates: VertexSet,
    mut excluded: VertexSet,
    clique_in_progress: &CliqueInProgress,
) where
    VertexSet: VertexSetLike,
    Accumulator: CliqueAccumulator,
{
    debug_assert!(candidates.all(|&v| graph.is_connected(v)));
    debug_assert!(excluded.all(|&v| graph.is_connected(v)));
    debug_assert!(candidates.is_disjoint(&excluded));
    debug_assert!(candidates.is_empty().not());

    while let Some(v) = candidates.pop_arbitrary() {
        let neighbours = graph.neighbours(v);
        let neighbouring_candidates: VertexSet =
            candidates.intersection(neighbours).copied().collect();
        if neighbouring_candidates.is_empty().not() {
            visit(
                graph,
                consumer,
                neighbouring_candidates,
                excluded.intersection(neighbours).copied().collect(),
                &clique_in_progress.pile(v),
            );
        } else if clique_in_progress.height + 1 >= consumer.min_clique_size
            && excluded.is_disjoint(neighbours)
        {
            let clique = clique_in_progress.pile(v);
            consumer.accept(clique.iter().copied().collect());
        }
        excluded.insert(v);
    }
}
